src/V3FuncOpt.cpp

```diff
@@ -0,0 +1,184 @@
+// -*- mode: C++; c-file-style: "cc-mode" -*-
+//*************************************************************************
+// DESCRIPTION: Verilator: Generic optimizations on a per function basis
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2003-2024 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+//
+// - Split assignments to wide locations with Concat on the RHS
+//   at word boundaries:
+//    foo = {l, r};
+//   becomes (recursively):
+//    foo[_:_] = r;
+//    foo[_:_] = l;
+//
+//*************************************************************************
+
+#include "V3PchAstMT.h"
+
+#include "V3FuncOpt.h"
+
+#include "V3Global.h"
+#include "V3Stats.h"
+#include "V3ThreadPool.h"
+
+VL_DEFINE_DEBUG_FUNCTIONS;
+
+class FuncOptVisitor final : public VNVisitor {
+    // NODE STATE
+    //  AstNodeAssign::user()     -> bool.  Already checked, safe to split. Omit expensive check.
+
+    // STATE - across all visitors
+    AstCFunc* const m_funcp;  // The function being processes
+
+    // STATE - Statistic tracking
+    VDouble0 m_concatSplits;  // Number of splits in assignments with Concat on RHS
+
+    // True for e.g.: foo = foo >> 1; or foo[foo[0]] = ...;
+    static bool readsLhs(AstNodeAssign* nodep) {
+        // It is expected that the number of vars written on the LHS is very small (should be 1).
+        std::vector<const AstVar*> lhsWrVarps;
+        std::unordered_set<const AstVar*> lhsRdVarps;
+        nodep->lhsp()->foreach([&](const AstVarRef* refp) {
+            if (refp->access().isWriteOrRW()) lhsWrVarps.emplace_back(refp->varp());
+            if (refp->access().isReadOrRW()) lhsRdVarps.emplace(refp->varp());
+        });
```

Comment:
I don't see why you need to build a vector, versus just checking each AstVarRef against refp->varp() and incrementing a count.

Reply:
1. I'm not sure the AstVars are local to the function, so I can't use userp (multiple CFucns are processed in parallel on the thread pool).
2. Because the expected size of lhsWrVarps is almost always 1, a vector is faster than a map of counters.

Reply:
I agree it's usually one, but if it's something very large as an edge case (and edge cases generally cause the performance problems), it seems inefficient.  Is an unordered_set really that much slower?  You're already using one for reads, and there will be more reads than writes.

Reply:
Or unordered_map in this case.

Reply:
> You're already using one for reads, and there will be more reads than writes.

`lhsRdVarps` needs to be looked up as many times as the *number of variables written on the LHS (generally 1)*, so a set is fine.

`lhsWrVarps` needs to be looked up as many times as the *number of variables read on the RHS (many)*, so a vector with a few elements can be a lot faster.

However, I went ahead and used a set, and special cased it for size 1, which will be good enough for the common case (1 written on the LHS), and won't blow up on a corner case (that I'm not sure can exists, but probably does :)).

> Is an unordered_set really that much slower

I keep coming back to stomping out unordered_foo to great effect over the years. Its O(1), but with a big 'c'. A loop over a small array is often faster.

Reply:
> I keep coming back to stomping out unordered_foo to great effect over the years. Its O(1), but with a big 'c'. A loop over a small array is often faster.

That's what I thought, but last time you used it, I thought you had convinced me that was no longer true!

_Originally posted by @gezalore in https://github.com/verilator/verilator/pull/5516#discussion_r1791747845_

> So has not been an issue for ~9 years. ...

Reply:
> That's what I thought, but last time you used it, I thought you had convinced me that was no longer true!

That's a different bad thing about sets :) (they used to be expensive to allocate even if empty/small). Here I'm just talking about the O(1) lookup cost.